### ISSUE TITLE
upstream connect error or disconnect/reset before headers. reset reason: protocol error

### ISSUE THREAD

**Issue by olizarevichroman** [dotnet/aspnetcore#43874]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We've launched ASP NET Core 7 app in a Kubernetes cluster with Envoy as a proxy.

When we send  `PUT` requests and a response contains a body - we have `502 upstream connect error or disconnect/reset before headers. reset reason: protocol error` for `PUT`. For `GET` it works, for `PUT` without a response body it also works.

After downgrading to `NET 6` it works

### Expected Behavior

A response should be `200/201`

### Steps To Reproduce

Send a `PUT` HTTP request to an endpoint that returns a result with a body.

### .NET Version

7.0.0-preview.7.22375.6

### Anything else?

ASP NET Core runtime version: `7.0.0-preview.7.22376.6`
NET Core runtime version: `7.0.0-preview.7.22375.6`

**Comment by Tratcher** [comment]
Please share the server logs for this request. 

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0#configure-logging

What was the response status code for 6.0? A 204? 204's aren't allowed to have a response body, and 7.0 enforces that.

**Comment by olizarevichroman** [comment]
@Tratcher 
I've found that we tried to return `ObjectResult` passing empty object as parameter with `204` status code.
So, after changing it with `NoContentResult` it works fine.

 But anyway, is it supposed to have more specific error for such cases? Getting `502 upstream connect error or disconnect/reset before headers. reset reason: protocol error` doesn't provide any useful details and tells nothing about `204` with body.

Thanks!

**Comment by olizarevichroman** [comment]
@Tratcher Also, what did you mean by `204's aren't allowed to have a response body, and 7.0 enforces that.`?

Do you know how does `7.0` enforce that?

**Comment by Tratcher** [comment]
Was that log from 6 or 7? I see a 204 with a Content-Length of 2. In 7.0 Kestrel will throw an exception if the status code is 204 and the app writes to the body. I don't see an exception in the logs, but I also don't see the 2 byte body being written.

**Comment by olizarevichroman** [comment]
@Tratcher 
yup, i have an exception in logs like [here](https://github.com/dotnet/aspnetcore/files/9550910/exception_log.txt)

But response still goes back without error) Tried it via Postman and this action, getting back 204. Is it expected?

```csharp
[HttpPut("update")]
public IActionResult Update()
{
    var value = new { };

    return new ObjectResult(value)
    {
        StatusCode = StatusCodes.Status204NoContent,
    };
}
```

**Comment by Tratcher** [comment]
That's not ideal. It should be changed to a 500 at least. I think @adityamandaleeka was still cleaning that up.

**Comment by olizarevichroman** [comment]
Sounds good) @Tratcher,  thanks a log)

**Comment by Tratcher** [comment]
This looks related to https://github.com/dotnet/aspnetcore/issues/43316#issuecomment-1223261301.
